### PR TITLE
DOC: Fix bad reference in ``numpy.ma``

### LINKED ISSUE #13114: DOC: missing references when using -n in sphinx-build
Started in #13104, turned into a tracking issue here.

Turning on -n in sphinx-build results in over 1000 warnings about bad references in docs. Trying to break these down with various grep statements, it seems about 200 are from numpy.ma, about 150 from polynomial and about 150 from c:type. That is under half of the WARNINGS.

~My workflow is to cd to the `doc` directory, modify the `ALLSPHINXOPTS` in the `Makefile and add  `-n`~, My workflow is to  set an environment variable SPHINXOPTS='-n' then run
```
(cd ..; pip install .); make clean; make html 2>&1 |tee /tmp/build.txt
less /tmp/build.txt
```
and look for the most common lines. Unfortunately this takes quite a while.

Anyone who wishes to contribute can issue PRs, preferably in chunks that address one class of failure. #13104 gives an idea of the changes needed.

### PR DESCRIPTION
**Pull request by takanori-pskq**
Partly fixes #13114

### CODE REVIEW

**Comment by charris** [comment]
close/reopen

**Comment by rossbar** [comment]
Quick question @takanori-pskq - you've marked this as WIP: does that mean that it is not ready for review in your opinion?

Also note that you don't have to fix *all* of the issues in #13114 in one big PR - in fact it would be preferable (from a review standpoint) if changes were broken up into smaller PRs that deal with coherent portions of the documentation. It seems this is how you originally intended to organize this PR (limiting changes to `numpy.ma`), but 9f38b49 introduces a large change to the scalar types docs.

**Comment by takanori-pskq** [comment]
@rossbar I marked as WIP because the build fails (that will be resolved by PRs #16296, #16311). And your comment about the commit 9f38b49 is right. I'll revert and create another PR.

**Comment by rossbar** [comment]
@takanori-pskq now that #16296 and #16311 are merged, do you want to try to rebase on master and see if that fixes the CI? I'd be happy to take a look afterwards!

File: numpy/core/_add_newdocs.py [hunk 1]
```diff
@@ -2322,7 +2322,7 @@
     See Also
     --------
     numpy.reshape : similar function
-    ndarray.reshape : similar method
+    numpy.ndarray.reshape : similar method
```
Comment (rossbar) [inline numpy/core/_add_newdocs.py:2325]: The original was okay here, see: https://numpy.org/devdocs/reference/generated/numpy.ndarray.shape.html

File: numpy/core/_add_newdocs.py [hunk 2]
```diff
@@ -4027,8 +4027,8 @@
 
     See Also
     --------
-    ndarray.T : Array property returning the array transposed.
-    ndarray.reshape : Give a new shape to an array without changing its data.
+    numpy.ndarray.T : Array property returning the array transposed.
+    numpy.ndarray.reshape : Give a new shape to an array without changing its data.
```
Comment (rossbar) [inline numpy/core/_add_newdocs.py:4031]: These changes aren't necessary either:

https://numpy.org/devdocs/reference/generated/numpy.ndarray.transpose.html

File: numpy/core/fromnumeric.py [hunk 3]
```diff
@@ -3141,8 +3141,8 @@ def size(a, axis=None):
     See Also
     --------
     shape : dimensions of array
-    ndarray.shape : dimensions of array
-    ndarray.size : number of elements in array
+    numpy.ndarray.shape : dimensions of array
+    numpy.ndarray.size : number of elements in array
```
Comment (rossbar) [inline numpy/core/fromnumeric.py:3145]: I assume these changes are to get the links in `np.ma.size` to work?

File: numpy/core/fromnumeric.py [hunk 4]
```diff
@@ -3562,8 +3562,8 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue):
 
     See Also
     --------
-    std, mean, nanmean, nanstd, nanvar
-    ufuncs-output-type
+    numpy.std, numpy.mean, numpy.nanmean, numpy.nanstd, numpy.nanvar
+    numpy.ufuncs-output-type
```
Comment (rossbar) [inline numpy/core/fromnumeric.py:3566]: Note that the `ufuncs-output-type` is a label defined in a document, so this doesn't work to resolve the link here.

File: numpy/core/fromnumeric.py [hunk 5]
```diff
@@ -3562,8 +3562,8 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue):
 
     See Also
     --------
-    std, mean, nanmean, nanstd, nanvar
-    ufuncs-output-type
+    numpy.std, numpy.mean, numpy.nanmean, numpy.nanstd, numpy.nanvar
```
Comment (rossbar) [inline numpy/core/fromnumeric.py:3565]: Here again, I assume the purpose is to get the links in `np.ma.var` to work correct? Note that there is already a `np.ma.std` and `np.ma.mean`, so those two links work appropriately (and target the ma versions of those functions, which is preferable).

However there aren't `ma` versions of the `nan*` functions, perhaps the approach works for those.

**Review by rossbar — COMMENTED** [review]
Thanks for taking a stab at this @takanori-pskq , this is certainly an improvement in fixing up the broken links in `np.ma`: I recorded 806 before this change (on 19f91f) and 738 afterwards. I think many of the changes are non-controversial. However, there is one pattern where `numpy.` is prepended to functions in the *numpy* docstrings so that links in the corresponding `numpy.ma` docstrings resolve to the numpy versions. This is okay, but has the unfortunate side-effect of cluttering up the original numpy docstrings a bit. Do you think there might be another way to get those links resolved using a different approach?

This PR touches a lot of different things so it will take me few passes before I can wrap my head around everything - I hope you can bear with me!

**Comment by takanori-pskq** [comment]
@rossbar Instead of just copying docstrings from numpy versions, how about writing the docstrings of the functions in `numpy.ma` like "This function is `numpy.ma` version of the function `foo`. See `numpy.foo` for details"?

**Comment by rossbar** [comment]
> Instead of just copying docstrings from numpy versions, how about writing the docstrings of the functions in numpy.ma like "This function is numpy.ma version of the function foo. See numpy.foo for details"?

I tend to agree in principle, but I don't think that this is a general solution. While some functions in `ma` simply use the docstring from their `np` counterpart , some don't. There are examples of both patterns (and others) in `numpy/ma/core.py` for example. I am not familiar enough with `np.ma` to say whether an approach like this would scale well.

Nevertheless, there are some changes in this PR that do indeed correct some broken links without affecting existing docstrings in other namespaces. Perhaps we could separate those out from the others that might need a more detailed discussion? If you're interested in doing so I'd be happy to help try to ID the straight-forward changes.

**Comment by takanori-pskq** [comment]
I'll separate the changes prepending `numpy.` (the lines you commented). Is there any changes that should be separated?

**Comment by rossbar** [comment]
> I'll separate the changes prepending numpy. (the lines you commented). Is there any changes that should be separated?

I think that's a good place to start. The addition of the missing functions to the `routines` doc and the corresponding `<module>.__all__` should certainly stay in.

**Review by rossbar — DISMISSED** [review]
LGTM - fixes about 25 broken xrefs in the ma docs (794 link warnings before, 769 after this change).

**Comment by mattip** [comment]
It seems this exposes new functions, should they have tests (maybe they already do)?

**Comment by rossbar** [comment]
> It seems this exposes new functions, should they have tests (maybe they already do)?

Good point - AFAICT, the related creation functions that are already exposed in np.ma (e.g. `np.ma.zeros`) aren't explicitly tested in `numpy/ma/tests`. Maybe this is an oversight that should be remedied? I'm not sure what the policy is for testing the ma functions that are derived from their non-masked counterparts (i.e. the functions like `ones_like` etc., which are defined in `numpy/ma/core.py` via the `_convert2ma` class).

**Comment by takanori-pskq** [comment]
rebased, and added tests for some functions.

File: numpy/ma/tests/test_core.py [hunk 6]
```diff
@@ -3186,6 +3186,96 @@ def test_empty(self):
         b = a.view(masked_array)
         assert_(np.may_share_memory(a.mask, b.mask))
 
+    def test_zeros(self):
```
Comment (rossbar) [inline numpy/ma/tests/test_core.py:3189]: The new tests here seem fine, but I would prefer either that the `x` and `x_like` tests were either broken up into to separate tests (e.g. `test_zeros` and `test_zeros_like`) or that the test names be amended to encompass both (e.g. `test_zeros_and_zeros_like`)

File: numpy/ma/tests/test_extras.py [hunk 7]
```diff
@@ -1686,3 +1686,103 @@ def test_stack_nd(self):
         assert_equal(c.shape, c_shp)
         assert_array_equal(a1.mask, c[..., 0].mask)
         assert_array_equal(a2.mask, c[..., 1].mask)
+
+class TestSplit:
```
Comment (rossbar) [inline numpy/ma/tests/test_extras.py:1690]: The tests here aren't specific to the `ma` versions of these functions. For example, they don't test that the output of `np.ma.split`, `np.ma.hsplit`, etc. are MaskedArrays. The non-ma functionality is already covered in the non-ma version of the tests (e.g. `numpy/lib/tests/test_shape_base.TestSplit`).

**Review by rossbar — COMMENTED** [review]
A couple comments about the tests that have been added, but everything else LGTM

**Comment by takanori-pskq** [comment]
rebased and fixed the tests.

**Comment by mattip** [comment]
There are a few comments open about shuffling tests around. @takanori-pskq would you like to relate to them or should we put this in as-is and somehow mark that for a later refactoring?

File: numpy/ma/tests/test_extras.py [hunk 8]
```diff
@@ -1686,3 +1686,106 @@ def test_stack_nd(self):
...
+    """Only testing for integer splits.
+
+    """
+    def test_0D_array(self):
+        a = masked_array(np.array(1), mask=[True])
+        try:
+            hsplit(a, 2)
+            assert_(0)
+        except ValueError:
+            pass
```
Comment (rossbar) [inline numpy/ma/tests/test_extras.py:1714]:
```suggestion
        try:
            assert_raises(ValueError, hsplit, a, 2)
```

File: numpy/ma/tests/test_extras.py [hunk 9]
```diff
@@ -1686,3 +1686,106 @@ def test_stack_nd(self):
...
+        a = masked_array(np.array(1), mask=[True])
+        assert_raises(ValueError, vsplit, a, 2)
+
+    def test_1D_array(self):
+        a = masked_array(np.array([1, 2, 3, 4]), mask=[1, 0, 1, 0])
+        try:
+            vsplit(a, 2)
+            assert_(0)
+        except ValueError:
+            pass
```
Comment (rossbar) [inline numpy/ma/tests/test_extras.py:1747]:
```suggestion
        assert_raises(ValueError, vsplit, a, 2)
```

File: numpy/ma/tests/test_extras.py [hunk 10]
```diff
@@ -1686,3 +1686,106 @@ def test_stack_nd(self):
...
+        assert_raises(ValueError, dsplit, a, 2)
+
+    def test_2D_array(self):
+        a = masked_array(np.array([[1, 2, 3, 4], [1, 2, 3, 4]]),
+                mask=[[0, 1, 0, 1], [1, 0, 1, 0]]);
+        try:
+            dsplit(a, 2)
+            assert_(0)
+        except ValueError:
+            pass
```
Comment (rossbar) [inline numpy/ma/tests/test_extras.py:1776]: Same pattern as above

**Review by rossbar — COMMENTED** [review]
A couple suggested cleanups to the tests but otherwise LGTM!

**Review by rossbar — DISMISSED** [review]
As it stands, the newly-exposed `zeros_like`, `ones_like`, and `full_like` functions won't support some of the masked array parameters (e.g. `fill_value`, `hardmask`). I think this is okay as `empty_like`, which is already exposed, doesn't either. I'm fine with following the precedent set by `empty_like` and leaving any potential updates to parameter support for the `*_like` arguments for a future PR.

Everything else LGTM, thanks @takanori-pskq for taking the time to re-organize the tests.

**Review by rossbar — COMMENTED** [review]
I was looking at this a bit more closely and just noticed an unfortunate name collision in the argument names of `np.ma.full`: `fill_value` is the name for *both* the value that `full` uses to fill the array, **and** the `fill_value` of the mask. In combination with `_convert2ma`, this can result in nasty errors:

```python
>>> np.ma.full(3, fill_value=5)
Traceback (most recent call last)
   ...
TypeError: full() missing 1 required positional argument: 'fill_value'
```

There might be a way around this using position-only arguments, but unfortunately I don't think the `_convert2ma` machinery is sufficient to handle this, at least not in it's current form.

**Comment by mattip** [comment]
Can we merge this PR as-is and continue cleanup afterwards or is there still work that needs to be done here? 

When we merge it, we should re-open gh-13114 since this is marked as closing it, or is there a way to tell github not to close it on merge

**Comment by eric-wieser** [comment]
I think @rossbar's concerns are valid - this PR mixes two things:

* DOC: Fixing references in docstrings
* ENH: Add new `numpy.ma.*` aliases for `numpy.*` functions

The second point needs more work, and looks problematic in its current form.

File: numpy/ma/core.py [hunk 11]
```diff
@@ -8123,14 +8123,16 @@ def __call__(self, *args, **params):
...
 identity = _convert2ma(
     'identity', params=dict(fill_value=None, hardmask=False))
 indices = np.indices
 ones = _convert2ma('ones', params=dict(fill_value=None, hardmask=False))
-ones_like = np.ones_like
+ones_like = _convert2ma('ones_like')
 squeeze = np.squeeze
 zeros = _convert2ma('zeros', params=dict(fill_value=None, hardmask=False))
-zeros_like = np.zeros_like
+zeros_like = _convert2ma('zeros_like')
```
Comment (eric-wieser) [inline numpy/ma/core.py:8135]: Can we split the python changes to a separate PR?

**Comment by BvB93** [comment]
FYI, the the new `np.ma.*` aliases have been moved to https://github.com/numpy/numpy/pull/17480.

**Review by rossbar — APPROVED** [review]
Looks good: 410 -> 397 warnings. Thanks @takanori-pskq !
